agent/grpc-internal/balancer/custombalancer.go

```diff
@@ -45,7 +45,7 @@ func (b *customPickfirstBalancer) UpdateClientConnState(state balancer.ClientCon
 		// of addresses matched the currently connected address, it would
 		// be an effective no-op.
 		if a.Equal(b.activeAddr) {
-			return nil
+			break
```

Comment:
Returning early would mean the rest of the addresses never get stored. Break instead.
This patch has been applied during the backports to 1.12.x (https://github.com/hashicorp/consul/pull/15834/commits/007a0f5b90d7f1c45c24ca61228288c3241f7f4d) and 1.13.x (https://github.com/hashicorp/consul/pull/15835/commits/5028b3f0bac8ab732e34b063210e8c2e7d555d87)